### ISSUE TITLE
Add a config singleton

### ISSUE THREAD

**Issue by liamhuber** [pyiron/pyiron_workflow#493]
# Why

While each node should track its inputs and outputs, it would be useful for node developers to be able to specify customizable rather than fixed defaults.

E.g., consider the following:

```python
import pyiron_workflow as pwf

@pwf.as_function_node
def RunFoo(executable="foo.exe", option=42):
    import subprocess
    subprocess.run([executable, "-fooopt", str(option)]) 
```

Then consider a user who does not have `foo.exe` in their path, but rather has their own custom complied version at `/some/path/to/my_foo.exe`.  It may be convenient for them to specify this custom default _once_ in a config setting, and then leave it be.

One can imagine a solution where the node developer writes something like

```python
import pyiron_workflow as pwf

pwf.config.set_if_missing("foo.exe", "my_awesome_node_package, "foo_executable")


@pwf.as_function_node
def RunFoo(executable=None, option=42):
    executable = pwf.config.my_awesome_node_package.foo_executable if executable is None else executable
    import subprocess
    subprocess.run([executable, "-fooopt", str(option)]) 
```

Then the user goes to some config file location (e.g. `~/pyiron_workflow.config`) and updates the toml (yaml? whatever) to read

```
[my_awesome_node_package]
foo_executable = "/some/path/to/my_foo.exe"
```

Now this user gets the updated default.

# Requirements

- Config should be optional -- nothing in the config should be absolutely required for the `pyiron_workflow` engine itself
- Config should be able to be static, e.g. it can be written in a human-readable file
- The file (if it exists) should be somewhere easy to find
- The location to look the file should _also_ be flexible (e.g. with a system env variable)
- Node developers should be able to easily provide defaults for configuring their packages that _don't_ over-ride 

Non-goals:
- Ensuring that two packages don't write to the same config fields (hard to stop, and anyhow largely avoidable by having the norm be using the node package conda name as its heading, as those are themselves guaranteed unique)
- Actually controlling `pyiron_workflow` itself with this config (although maybe one day in the future we want to)

# Mock solution

We already do this in `pyiron_base`, so we can lean heavily on what's there.
Basically you just need the engine to instantiate an object like this:

```python
import typing

from pyiron_snippets.singleton import Singleton
from pyiron_snippets.dot_dict import DotDict

class Config(DotDict, metaclass=Singleton):
    def __init__(self):
        """
        First try reading from a system env location, like PYIRON_WORKFLOW_CONFIG
        Next, try reading from a fixed location, like `~/.pyiron_workflow.config`
        
        Finally, add the location from which the config was read to the configuration itself
        (for debugging. Just `None` if there was no file at the env or default locations).
        """
        raise NotImplementedError

    def set_if_missing(value: typing.Any, *toml_path: str):
        """
        Interpret the toml path (or yaml, or whatever format gets used) by concatenating the args,
        but also correctly interpret if it's all passed "as_one.dotted.path_to.a_subvariable".
        If the value is already present, don't do anything, otherwise set it.

        Maybe there is a nicer name for this method?
        """
```

And then in `pyiron_workflow.__init__` we initialize such a singleton.

**Comment by XzzX** [comment]
This will make nodes non self-contained anymore! This complicates reproducability and save/load!

Why singleton? `pwf.config` does not look like a singleton.

Please avoid global variables whenever possible.

**Comment by liamhuber** [comment]
> This will make nodes non self-contained anymore! This complicates reproducability and save/load!

> Please avoid global variables whenever possible.

Yep, convincing arguments to me. This sort of behaviour was explicitly requested in the weekly meeting at some point. @ligerzero-ai, I guess stuff like the VASP executable path is what was in mind here. How do you feel about @XzzX's arguments? If you can live without this, I'd be happy to close the issue without implementing this.

The other place I can see this sort of thign being useful is for configuring stuff like `pysqa` queues; but hopefully `pysqa` itself (or `executorlib` or whomever) can handle this themselves and we don't need to make it our responsibility here.

> Why singleton? 

If the user updates the config manually in their python session, and then some other part of their code tries to reference the config, we want that reference to get the same object and detect the changes.

> `pwf.config` does not look like a singleton.

Unclear; what should a singleton "look" like?

**Comment by ligerzero-ai** [comment]
**I have no strong opinions either way.**

Nodes will never be self contained because of they necessarily call external executables in scientific workflows - we don't have control over those and never will. We can't wrap an entire executable and ship it off with every single instance of the node - we're not in the business of perfect reproducibility, but pragmatic reproducibility.

The question is how we want users to interact with nodes, do we want to make them specify the executable path every single time they run a workflow?

If this is not implemented, I am likely going to hack this in on the user side, anyway. 

A reference to call "/cmmc/u/ptmp/hmai/VASP/vasp.exe" in a shell node is not going to be reproducible on other machines, anyway.

The other side of the argument is that if everyone eventually does this, I don't see why it would pose a problem?

Probably a topic for Jorg to chime in on, which side of the fence we want to live on. Classical pyiron is about user-convenience and abstracting this technical config stuff away from the user.

**Comment by XzzX** [comment]
> Nodes will never be self contained because of they necessarily call external executables in scientific workflows - we don't have control over those and never will. We can't wrap an entire executable and ship it off with every single instance of the node - we're not in the business of perfect reproducibility, but pragmatic reproducibility.

Yes, but we have to give at least some promises. Imagine the following situation: A user writes their own MD node with a Lennard Jones potential. However, he does not pass `sigma` and `epsilon` via input parameters but via a global config. How do you detect changes? When do you rerun the node? How to save it, so it can be loaded in another notebook instance? This breaks one of the fundamental promises of functional programming I thought we wanted to take advantage of, namely same input -> same output.

My suggestion would be. Provide a config node (dataclass) and pass it as input. Might complicate the graph significantly if a lot of small config values need to be inputed. Therefore, we could also think about passing a complete dict as an input, and the node takes what it needs. This way it makes it trackable and storable.

> 
> The question is how we want users to interact with nodes, do we want to make them specify the executable path every single time they run a workflow?
> 
> If this is not implemented, I am likely going to hack this in on the user side, anyway.

You are free to do so. The question is, do we provide it as library authors and give some promises/guarantees how it works or do you do it as a end-user. In the first case we are responsible. In the second case the user is responsible for unwanted side effects.

**Comment by niklassiemer** [comment]
I would be possible to have a config dataclass at the workflow level and each node not having a config as input gets that one, i.e. it would be a default connection from the workflow.config dataclass to the always present input.config channel. Thus, there would be a workflow 'global' config dataclass used as input. The feature of not having to care about this input channel would be retained without the need to construct an implicit global Singleton. Would this be a possible common ground?

**Comment by XzzX** [comment]
> I would be possible to have a config dataclass at the workflow level and each node not having a config as input gets that one, i.e. it would be a default connection from the workflow.config dataclass to the always present input.config channel. Thus, there would be a workflow 'global' config dataclass used as input. The feature of not having to care about this input channel would be retained without the need to construct an implicit global Singleton. Would this be a possible common ground?

That is very much along the lines what I have suggested. So, from my side, yes.

Remark: The "hash-based database" will then store the complete graph with every config change.

**Comment by liamhuber** [comment]
Pre-`flowrep` issue
